### PR TITLE
Multi-frame shader compilation

### PR DESCRIPTION
**Pull request by wdanilo**
[ci no changelog needed]

### Pull Request Description

Implements [#181681597](https://www.pivotaltracker.com/story/show/181681597). This is a PR that uses a lot of code from https://github.com/enso-org/enso/pull/3372. 

image

### Important Notes

- This code has a `FIXME` left in 2 places. These are small refactorings – that will be fixed after the first review.
- This code exposes [tracing](https://crates.io/crates/tracing) crate in prelude. We will be moving to `tracing` from our logger in the near future, but `tracing_wasm` needs to be fixed to support our use cases (@mwu-tow is working on it).
- The application needs to be heavily tested as this code touches core rendering concepts.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

### CODE REVIEW

File: lib/rust/ensogl/core/src/display/symbol/gpu.rs [hunk 1]
```diff
@@ -419,8 +419,13 @@ impl Symbol {
                 }
                 if self.shader_dirty.check() {
                     let var_bindings = self.discover_variable_bindings(global_variables);
-                    self.shader.update(&var_bindings);
-                    self.init_variable_bindings(&var_bindings, global_variables);
+                    // FIXME: Potential mem leak
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/symbol/gpu.rs]: This will be fixed

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+                };
+
+                let vertex = &shader.vertex;
+                let fragment = &shader.fragment;
+                let vertex_log = context.blocking_format_error_log(vertex);
+                let fragment_log = context.blocking_format_error_log(fragment);
+                let vertex_error = unwrap_error("Vertex", vertex_log);
+                let fragment_error = unwrap_error("Fragment", fragment_log);
+
+                // FIXME: this should be taken from config and refactored to a function
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:438]: this will be fixed

File: lib/rust/ensogl/core/src/animation/loops.rs [hunk 3]
```diff
@@ -14,25 +14,29 @@ use web::Closure;
...
 #[derive(Clone, Copy, Debug, Default)]
+#[allow(missing_docs)]
 pub struct TimeInfo {
-    /// Start time of the animation loop.
-    pub start: f32,
-    /// The last frame time.
-    pub frame: f32,
-    /// The time which passed since the animation loop was started.
-    pub local: f32,
+    pub animation_loop_start:         f32,
```
Comment (MichaelMauderer) [inline lib/rust/ensogl/core/src/animation/loops.rs]: Might be good to also add information about the time units that are used here.

File: lib/rust/ensogl/core/src/animation/loops.rs [hunk 4]
```diff
@@ -95,7 +99,8 @@ impl<Callback> RawLoopData<Callback> {
         Self { callback, on_frame, handle_id }
     }
 
-    /// Run the animation frame.
+    /// Run the animation frame. The time fill be converted to [`f32`] because of performance
```
Comment (MichaelMauderer) [inline lib/rust/ensogl/core/src/animation/loops.rs]: `fill` -> `will`

File: lib/rust/ensogl/core/src/animation/loops.rs [hunk 5]
```diff
@@ -95,7 +99,8 @@ impl<Callback> RawLoopData<Callback> {
...
+    /// reasons. See https://hugotunius.se/2017/12/04/rust-f64-vs-f32.html to learn more.
```
Comment (MichaelMauderer) [inline lib/rust/ensogl/core/src/animation/loops.rs:104]: I remember we had at some point issues with precision of long-running animations. We should ensure this does not introduce an issue with that again.

File: lib/rust/ensogl/core/src/system/gpu/context/extension.rs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+/// [https://www.khronos.org/registry/webgl/extensions/KHR_parallel_shader_compile]
+#[derive(Clone, Copy, Debug)]
+#[allow(missing_docs)]
+pub struct KhrParallelShaderCompile {
+    pub completion_status_khr: GlEnum,
+}
+
+impl KhrParallelShaderCompile {
+    /// Try to obtain the extension.
+    pub fn init(context: &WebGl2RenderingContext) -> Option<Self> {
```
Comment (MichaelMauderer) [inline lib/rust/ensogl/core/src/system/gpu/context/extension.rs]: Could this be called `try_init`?

File: lib/rust/ensogl/core/src/system/gpu/context/extension.rs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    /// Asynchronously check if the job is ready.
+    pub fn is_ready(&self, context: &WebGl2RenderingContext, program: &WebGlProgram) -> bool {
+        let param = self.completion_status_khr;
+        context.get_program_parameter(program, *param).as_bool().unwrap_or_else(|| {
+            REPORTABLE_WARNING!(
+                "context.getProgramParameter returned non bool value for KHR \
+                Parallel Shader Compile status check. This should never happen, however, it should \
+                not cause visual artifacts. Reverting to non-parallel mode."
```
Comment (MichaelMauderer) [inline lib/rust/ensogl/core/src/system/gpu/context/extension.rs]: The ` Reverting to non-parallel mode.` is not actually implemented in this method, though. So, I'm wondering whether this info should be emitted where it is actually done to avoid future refactoring breaking this warning message.

File: lib/rust/ensogl/core/src/system/gpu/context/native.rs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+// === BlockingCheckStatus ===
+// ===========================
+
+/// Status check for the given WebGL entity. For shaders, it checks the compilation status, for
+/// shader programs, the linking status. This operation is blocking and should be used only when
+/// necessary. To learn more, see:
+/// [https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/WebGL_best_practices#compile_shaders_and_link_programs_in_parallel]
+#[allow(missing_docs)]
+pub trait BlockingCheckStatus<T> {
+    fn blocking_check_status(&self, target: &T) -> Option<String>;
```
Comment (MichaelMauderer) [inline lib/rust/ensogl/core/src/system/gpu/context/native.rs]: It might be good to explain a bit more about the expected return types. `String` seems surprising at first. 

Same for the other return types in the trait and functions below.

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+
+// =================
+// === Constants ===
+// =================
+
+/// We do not want the framerate to drop below this value when compiling shaders. Whenever we
+/// discover that it drops below this threshold, no more shaders will be scheduled for compilation
+/// or linking in this frame. It does not mean, however, that the framerate will not be lower. The
+/// last scheduled shader can take a lot of time to be rendered, causing the FPS to drop
+/// significantly.
```
Comment (MichaelMauderer) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs]: Does this mean if for whatever reason the application is, e.g., limited to 30FPS, no shaders will ever be compiled? That would be bad.

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+
+    /// Get weak reference to this handler.
+    pub fn downgrade(&self) -> WeakJobHandler {
+        let weak = Rc::downgrade(&self.rc);
+        WeakJobHandler { weak }
+    }
+}
+
+impl WeakJobHandler {
+    /// heck whether the handler was dropped.
```
Comment (MichaelMauderer) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs]: `Check`

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+
+/// Compiles and links GL shader programs, asynchronously. The compiler works in the following way:
+/// 1. A new shader code is submitted with the [`submit`] method. A new job is created in the
+///    [`Jobs::compile`] queue and the job handler is returned to the user.
+///
+/// 2. The following pseudo-algorithm is performed:
+/// ```text
+/// on_ever_frame if job queues are not empty {
+///     Promote ready jobs from [`Jobs::khr_completion_check`] to [`Jobs::link_check`].
+///     Loop while the current frame time is < MAX_FRAME_TIME_MS {
```
Comment (MichaelMauderer) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs]: Repeat question of the comment from above: so on a weak system, we might never compile any shaders?

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+                    self.dirty = false;
+                },
+        }
+        Ok(())
+    }
+
+    #[profile(Detail)]
+    fn run_khr_completion_check_jobs(&mut self) {
+        debug!(self.logger, "Running KHR parallel shader compilation check job.");
+        let khr = self.context.extensions.khr_parallel_shader_compile.as_ref().unwrap();
```
Comment (MichaelMauderer) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs]: The `unwrap` should be an `except` with explanation.

File: lib/rust/ensogl/core/src/animation/loops.rs (hunk 5 above)
Comment (kazcw) [reply lib/rust/ensogl/core/src/animation/loops.rs:104]: Yeah, the precision issue @MichaelMauderer mentions seems problematic now that we are relying on these values in compiler logic. After half an hour an `f32` will reach half the effective precision of Chrome's `performance.now` values; at 1 hour, it will have 1/4 the precision, and so on.

There's a simple fix though: We have no reason not to use a f64 for this. The performance effect observed in the blog post linked above is not relevant here. The author of the post misidentified the cause of the phenomenon; almost all of the difference he observed is caused by autovectorization.

On x86-64 (without target_feature=avx2), the compiler backend implements floating point math with 128-bit vector operations. When floating point operations are performed on triplets of scalars like his benchmark, the compiler can perform all 3 operations with one instruction (because a 128-bit register holds up to 4 f32 values), but for f64 values it needs two instructions. Because his code is throughput-bound on floating point computation, that factor-of-2 difference in floating point instructions causes a large impact.

What we're doing is different; we only ever operate on one of these values at a time, so the compiler (in this case referring to the whole stack of llvm + the wasm runtime's JIT) has no opportunity for vectorization. It will take the same number of instructions whether we use f32 or f64. The other differences between single- and double-precision ops are not usually important, and in our case negligable because we won't be bottlenecking on the affected resources.

File: lib/rust/ensogl/core/src/display/scene.rs [hunk 13]
```diff
@@ -1086,6 +1085,7 @@ impl Scene {
                 self.update_shape();
                 self.update_symbols();
                 self.handle_mouse_over_and_out_events();
+                context.shader_compiler.run(time);
```
Comment (kazcw) [inline lib/rust/ensogl/core/src/display/scene.rs:1102]: In this implementation, we do these operations in a frame:
- Render.
- Check compilation-completions and start new compilations.

If we instead:
- Check compilation-completions.
- Render.
- Start new compilations.

Then we would be able to start using each shader 1 frame sooner in most cases. This would be noticeable mainly if the frame in question is a large missed frame.

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+            while self.dirty {
+                match self.run_step() {
+                    Ok(_) => {
+                        if !self.dirty {
+                            break;
+                        }
+                        let now = self.performance.now() as f32;
+                        let current_frame_time =
+                            now - time.animation_loop_start - time.since_animation_loop_started;
+                        if current_frame_time > MAX_FRAME_TIME_MS {
```
Comment (kazcw) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs]: This condition won't be true until we've already missed the frame, right?

File: lib/rust/ensogl/core/src/display/symbol/gpu/shader.rs [hunk 15]
```diff
@@ -145,17 +149,15 @@ impl {
...
-                    }
+                    let code = shader_builder.build();
 
+                    *self.program.borrow_mut() = None;
+                    let program = self.program.clone_ref();
+                    let handler = context.shader_compiler.submit(code, move |p| {
+                        *program.borrow_mut() = Some(p);
+                        on_ready();
+                    });
+                    self.shader_compiler_job = Some(handler);
```
Comment (kazcw) [inline lib/rust/ensogl/core/src/display/symbol/gpu/shader.rs]: It's a bit subtle that dropping the old handle here prevents a race condition. Could use a comment.

**Review by kazcw — COMMENTED** [review]
Besides my comments on the changeset, there's a performance issue. I just found while profiling my PR (this PR is also affected) that sometimes a COMPLETION_STATUS_KHR query will block for a long time (up to 500ms), probably until the whole pipeline flushes. I have an idea of what kind of chromium quirk is probably causing this, and I'm investigating now.

I think the changes necessary to fix it will be small, and even when the worst case behavior is hit this PR is a performance improvement, so the issue doesn't need to block progress on this PR.

**Comment by kazcw** [comment]
Details on the performance issue: It seems Chromium (and Electron) has limits on the number of compile/link jobs that can be in flight. If the limits are exceeded, it applies a crude form of backpressure by blocking until all pending jobs complete. We can avoid that happening by implementing limits on the number of jobs in the compiling and linking states. I've found that these (very low) limits work on my machine: ensure concurrent compiling jobs <= 2 (i.e. 2 pairs of shaders), ensure concurrent linking jobs <= 4. I expect those values would avoid getting throttled in any Chromium/Electron environment where khr_completion_status is supported, because chromium always forks at least 4 shader-compiling processes when the extension is available. I don't know the optimal values for these limits, or whether they're machine-dependent; it would be nice to find out how high we can safely raise the limits so we can get shaders compiled faster, but it isn't hugely important now that we aren't blocking on shader compilation.

**Comment by wdanilo** [comment]
@kazcw thank you so much for such a detailed check on these. We can either hardcode the max concurrent linking jobs or we can just spawn them until one blocks for longer and then remember the number for the rest of the app lifetime. What do you think about it?

**Comment by kazcw** [comment]
@wdanilo

Increasing parallelism would only benefit CPUs with more than 8 threads (due to the way Chromium determines its thread limits). Users with lower-end hardware would also be the most affected by hitting the blocking case. I wouldn't like to see low-end performance slowed down that much to improve high-end performance.

It would be useful to know what we'd gain by increasing parallelism (for hardware that supports it); I also really don't like not knowing for sure what's causing this. So I gathered more data: I used low-level profilers to change how we measure this code, so that we track not just what compile-stages we're executing, but how they fit into the lifetime of each shader. (I'll PR the profiling changes to your branch, but I need to separate it from my quick and dirty throttling implementation first.)

# Summary of findings

- The vast majority of the time we spend compiling shaders is when we initially render the graph; we don't do much compiling after coloring information arrives. So at least until the backend provides type info much sooner, increasing compile parallelism would not significantly impact total startup time.
- Some shader is taking 500ms to compile. This is probably related to the blocking case. I wonder if it has anything to do with the MacOS crashes? We need to find out what's going on with this shader.

# The data

## Without the job-limit workaround

Here's a profile without implementing limits, demonstrating the blocking case (rendered with #3382):
image

The top part shows the main thread; you can see where we block for about 500 ms on the completion check. Each line in the tall stack below is a shader compile job. All the shaders jobs finish at the same time, confirming that it is a full pipeline flush: The completion check blocks until everything is done compiling. (The timings of everything else in the main thread would normally be much shorter, but this is a debug/opt-level=1 build; optimization level doesn't affect the timings of the relevant calls.)

## With the workaround

Here's a proflie with a limit of 4 concurrent jobs--scale is about the same.
image

(I'm only showing the shader-compiling part here; compiling never blocks the main thread in this case.) The weird thing here is that first compile--it takes 500ms for one shader?? The shader in question is the first one to start compiling when we start preparing to draw the graph editor. I'll add some more instrumentation and find out what code that would be.

File: lib/rust/ensogl/core/src/system/gpu/context/extension.rs (hunk 6 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/system/gpu/context/extension.rs]: Sure, a way better name, thanks!

File: lib/rust/ensogl/core/src/system/gpu/context/extension.rs (hunk 7 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/system/gpu/context/extension.rs]: This is a very good comment. Moved this thing to the right place. Thanks for spotting it.

File: lib/rust/ensogl/core/src/system/gpu/context/native.rs (hunk 8 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/system/gpu/context/native.rs]: I renamed it to `blocking_get_status_logs` and added additional explanation in the docs.

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs (hunk 9 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs]: No, at least one shader will be processed per frame. Added explanation there.

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs (hunk 10 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs]: Heck, good catch.

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs (hunk 11 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs]: Good catch. The docs did not explain it properly. fixed the docs.

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs (hunk 12 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs]: You are right, this is implemented in a bad way. I implemented it without this line at all now.

File: lib/rust/ensogl/core/src/display/scene.rs (hunk 13 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/display/scene.rs:1102]: I did it this way on purpose. I see multiple problems with the approach proposed in your comment:
1. According to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/WebGL_best_practices#avoid_blocking_api_calls_in_production):
    image
    So, it might not be a good idea to evaluate `get*Parameter` functions on the beginning of a frame. 
2. Your `link_check` queue can be big in the compiler. Every `get*Parameter` is costly. You do not know on the beginning of the frame how much time you have left for these operations in this frame, because you don't know how much non-compiler computations are required.
3. This will make the shader compiler pipeline way more complex, while getting results 1 frame faster does not seem like a big win, especially as we are rendering things ~60 frames per second.

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs (hunk 14 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs]: That's right, but IMO we should be ok with it. I added additional explanation to the `FPS_THRESHOLD` constant. I also renamed this constant to `FRAME_TIME_THRESHOLD`, because `MAX_FRAME_TIME` was confusing indeed.

File: lib/rust/ensogl/core/src/display/symbol/gpu/shader.rs (hunk 15 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/display/symbol/gpu/shader.rs]: You are very right. Thanks for catching it. I made it more explicit:
image

**Comment by wdanilo** [comment]
@kazcw @MichaelMauderer I applied your comments here. It's ready for review and hopefully for merge. A few notes:

1. @kazcw, I created a limit for max parallel compilation jobs. However, as we do not know when the compilation is finished and when the linking starts, we set the limit globally, in such a way that there are max 2 compilation & linking jobs in total. Please check it out and if you'd have a better idea, let me know.
2. The `f32` vs `f64` time issue is important, but let's fix it in a separate PR. I was trying doing it in this PR, but it is not easy. After changing that, we need to either change the animation system to use `f64` everywhere, or generalise it. I don't think we want to use it everywhere, as it is used to animate colors, and `f32` is good enough for color representation. Moreover, `f32` could (should?) make the color management faster, as `sRGBA` stores just 4 `f32` side by side. Anyway, this is more complex topic for a separate PR.
3. Please test this PR and check if everything works in the product as it should. Then, let's merge it :)

File: lib/rust/ensogl/core/src/display/scene.rs (hunk 13 above)
Comment (kazcw) [reply lib/rust/ensogl/core/src/display/scene.rs:1102]: MDN is speaking very generally here. Most uses of `get*Parameter` have to flush and block in order to accurately return a value; `COMPLETION_STATUS_KHR` is an exception because it is a specifically non-blocking API, that returns whatever result is currently available. It's not free (it is a FFI call that corresponds directly to 1-2 GL calls), but per call it is microsecond-scale; we are spending <2ms total in these queries, spread out across many frames during startup.

However the benefit of doing this earlier would only be noticeable for big missed frames, so if it would complicate the implementation a lot, it is not worth it; we will get most of the same improvement if we break up functions that run after first render and take a lot of time at once, to improve responsiveness generally.

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 16]
```diff
@@ -0,0 +1,466 @@
...
+/// this threshold. Every frame at least one job is scheduled, and the last scheduled job can take
+/// significant amount of time, causing the FPS to drop below this threshold. To learn more about
+/// how jobs are scheduled, read the docs of the [`Compiler`].
+const FPS_THRESHOLD: f32 = 60.0;
+const FRAME_TIME_THRESHOLD: Duration = (1000.0 / FPS_THRESHOLD).ms();
+
+/// Maximum number of parallel shader compilation jobs. Chromium (and Electron) has limits on the
+/// number of compile/link jobs that can be in flight. If the limits are exceeded, it applies a
+/// crude form of backpressure by blocking until all pending jobs complete. To learn more, see:
+/// https://github.com/enso-org/enso/pull/3378#issuecomment-1090958946
```
Comment (kazcw) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:54]: I am not confident in this explanation, especially after gathering data and finding that the issue correlates to an outlier shader compilation. All I can say with confidence is that we tend to encounter some bug/feature that causes major jank during startup if we don't put a limit on parallelism. I would like to look in to it further but I don't think it needs to be done before this PR is merged.

File: lib/rust/ensogl/core/src/display/shape/primitive/system.rs [hunk 17]
```diff
@@ -76,7 +76,7 @@ impl ShapeSystem {
         material.add_input("pixel_ratio", 1.0);
         material.add_input("z_zoom_1", 1.0);
         material.add_input("time", 0.0);
-        material.add_input("symbol_id", 0);
+        // material.add_input("symbol_id", 0);
```
Comment (kazcw) [inline lib/rust/ensogl/core/src/display/shape/primitive/system.rs]: Why is this commented?

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 18]
```diff
@@ -0,0 +1,466 @@
...
+            debug!(self.logger, "{msg1} {msg2}");
+            no_progress
+        }
+    }
+
+    /// Get the number of shader compilation jobs run in parallel. Please note, that it is
+    /// impossible to get separate values for compilation and linking jobs here, because checking
+    /// it is costly and can prevent the parallelism altogether. To learn more, see:
+    /// https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/WebGL_best_practices#dont_check_shader_compile_status_unless_linking_fails
+    fn current_parallel_job_count(&self) -> usize {
```
Comment (kazcw) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:294]: Doesn't include compiling state. `+ link.len()`?

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 19]
```diff
@@ -0,0 +1,466 @@
...
+
+    /// Runs the next compiler job if there is any left and if it will not cause too many jobs being
+    /// run in parallel. The result [`bool`] indicates if the call to this function did any
+    /// progress.
+    #[profile(Detail)]
+    fn run_step(&mut self) -> Result<bool, Error> {
+        let did_progress = |_| Ok(true);
+        let no_progress = Ok(false);
+        let jobs = &self.jobs;
+        if self.current_parallel_job_count() < MAX_PARALLEL_COMPILE_JOBS {
```
Comment (kazcw) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs]: We could gate only `run_next_compile_job`, since none of the others increase the number of outstanding jobs (after changing `current_parallel_job_count`)

**Review by kazcw — APPROVED** [review]

File: lib/rust/ensogl/core/src/display/shape/primitive/system.rs (hunk 17 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/display/shape/primitive/system.rs]: It's not needed anymore, we are not using `symbol_id` in shaders. should be removed!

File: lib/rust/ensogl/core/src/display/scene.rs (hunk 13 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/display/scene.rs:1102]: You are right, the problem here is not `get*Parameter` for `COMPLETION_STATUS_KHR`, but there are several problems that require better description:
- On platforms where KHR extension is not accessible, we would need to use the blocking `get*Parameter` (the compilation can still be performed in parallel by the browser).
- Even if KHR is accessible, we still need to check linking status for each shader, which is costly operation and I think it requires flush + round-trip.
- Even if KHR is accessible, starting new compilations on the beginning of the frame is not a goo idea IMO, because starting the compilation takes some time as well. This actually can be caused by the fact that the shader code in Rust is kept in big Strings that need to be converted to JS and send to the driver, but I'm not sure about that. Anyway, we do not know how many compilations we can spawn at the beginning of the frame.

To sum this up, I see several problems and the solution is not obvious to me. Feel free to help me resolve them and if you see a solution, I'd be happy to implement an improvement here. On the other hand, this will probably make the implementation way more complex and the difference will probably never be spot on in real world cases (as you've written, there are other solutions for the problem of big missed frames).

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs (hunk 16 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:54]: Gotcha. Changing docs to be more subtle.

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs (hunk 18 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:294]: Hmmmm, that's right. I'm wondering how I missed that. Thanks

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs (hunk 19 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs]: Actually I think the current implementation can have strange corner cases and gating `run_next_compile_job` only is a way more bulletproof idea. Implemented.

File: lib/rust/types/src/unit2.rs [hunk 20]
```diff
@@ -0,0 +1,631 @@
+//! Utilities for creating custom strongly typed units. For example, unit `Duration` is a wrapper
```
Comment (MichaelMauderer) [inline lib/rust/types/src/unit2.rs:1]: I’m wondering whether we would not have been better off to use a pre-existing solution than re-implementing this ourselves and now having two different implementations of this in our code base. 

I'm not going to argue against including this as it is already written, but I think it might have been easier to use something off the shelves (e.g. `uom` or `chrono`) for our use case, which would lead to a smaller codebase on our side, making future maintenance easier for us.
 We should also strongly consider planning a chore to remove the old one (or both) in the midterm.

**Review by MichaelMauderer — APPROVED** [review]

File: lib/rust/types/src/unit2.rs (hunk 20 above)
Comment (wdanilo) [reply lib/rust/types/src/unit2.rs:1]: We have many units, including pixels, or some shader units. There is no package I know that provides. generic way of working with units. Also `unit2` is way better than `unit`, so I want to slowly migrate to it and remove the old `unit` over time.

Comment (wdanilo) [reply lib/rust/types/src/unit2.rs:1]: Moreover, how we keep the data under the hood is important. For example, if you want to keep time data in general, it would rather be a bad idea to just use `f64`. However, for our purposes this is the best representation ever.
